client/src/push-content.ts

```diff
@@ -207,6 +210,44 @@ export const getMessage = async (): Promise<string | undefined> => {
   return message
 }
 
+// NOTE: When the private submodule is not initialized, this will return undefined
+export const _getPrivateSubmodule = () => getRepo(PRIVATE_SUBMODULE_NAME)
+
+/* istanbul ignore next (already tested in pushContent) */
+export const initPrivateSubmodule = async (hostContext: ExtensionHostContext) => {
+  const privateSubmodule = _getPrivateSubmodule()
+  if (privateSubmodule !== undefined) {
+    await preparePrivateSubmodule(hostContext.client, privateSubmodule)
+  } else {
+    console.warn(`Private submodule not found: ${PRIVATE_SUBMODULE_NAME}`)
+  }
+}
+
+export const preparePrivateSubmodule = async (
+  client: LanguageClient,
+  privateSubmodule: Repository
+): Promise<void> => {
+  // TODO: What if submodule needs to be initialized? (not a problem in gitpod)
```

Comment:
I'm probably just confused but doesn't this code handle this `TODO`?

Reply:
No, initially when clone a repository that contains submodules, the submodules are not initialized. They do not contain a `.git` directory until you run something like `git submodule update --init`. Since they do not contain a `.git` directory, they are not considered git repositories and git commands like `fetch` fail.

Gitpod automatically initializes submodules when you open a workspace. This TODO is specifically for making it work locally. I was exploring ways to make it work. One possible way to get it to work right now would be to use the `git.clone` command. Since it works in Gitpod and you can workaround the issue locally with by initializing submodules manually, it seemed like something that could wait until later.

Reply:
I tried making this more clear by adding an error that is thrown if the private submodule is unexpectedly missing. This would happen when the submodule is defined in the `.gitmodules` and:
1. The directory is missing
2. The directory is empty
3. Any other situation where the git extension does not think it is a valid git repository